setoff_workflows/setoff_workflows.py

```diff
@@ -0,0 +1,2622 @@
+                "permissions": "VIEW",
+            },
+            "admins": {
+                "user_list": SWConfig.DNANEXUS_USERS["admins"],
+                "permissions": "ADMINISTER",
+            },
+        }
+
+    def write_project_creation_script(self) -> None:
+        """
+        Write the script that creates the DNAnexus project and shares it with the
+        required users with the required access levels. The project is created using
+        the project creation command which utilises the DNAnexus sdk. This command and
+        the project sharing commands are written to a bash script
+            :return None:
+        """
+        lines_to_write = [
+            SWConfig.SDK_SOURCE,
+            SWConfig.DX_CMDS["create_proj"]
+            % (
+                SWConfig.PROD_ORGANISATION,
+                self.samples_obj.nexus_paths["proj_name"],
+                self.dnanexus_auth,
+            ),
+        ]
+        # Give view and admin permissions for project
+        for permissions_level in self.users_dict.keys():
+            if self.users_dict[permissions_level]["user_list"]:
+                for user in self.users_dict[permissions_level]["user_list"]:
+                    lines_to_write.append(
+                        SWConfig.DX_CMDS["invite_user"]
+                        % (
+                            user,
+                            self.users_dict[permissions_level]["permissions"],
+                            self.dnanexus_auth,
+                        )
+                    )
+            else:
+                self.rf_obj.rf_loggers.sw.info(
+                    self.rf_obj.rf_loggers.sw.log_msgs["no_users"],
+                    permissions_level,
+                )
+        lines_to_write.append("echo $PROJECT_ID")
+        write_lines(self.rf_obj.proj_creation_script, "w", lines_to_write)
+
+    def run_project_creation_script(self) -> str:
+        """
+        Set off the project creation script using subprocess. The output of this command is checked to
+        ensure it meets the expected success pattern. If unsuccessful, exit script
+            :return projectid (str):    Project ID of the created project
+        """
+        self.rf_obj.rf_loggers.sw.info(
+            self.rf_obj.rf_loggers.sw.log_msgs["creating_proj"],
+            self.rf_obj.proj_creation_script,
+        )
+        project_creation_cmd = f"bash {self.rf_obj.proj_creation_script}"
+
+        project_id, err, returncode = execute_subprocess_command(
+            project_creation_cmd, self.rf_obj.rf_loggers.sw, "exit_on_fail"
+        )
+        if returncode == 0:
+            return project_id
+        else:
+            self.rf_obj.rf_loggers.sw.error(
+                self.rf_obj.rf_loggers.sw.log_msgs["proj_creation_fail"],
+                self.samples_obj.nexus_paths["proj_name"],
+                err,
+            )
+            sys.exit(1)
+
+    def get_upload_cmds(self) -> dict:
+        """
```

Comment:
Hey, the '%' is used because it is inputting a variable into a string with an s string that has been defined already in the config. f strings can only be used where the string and the variable to insert are defined in the same place